staging/src/k8s.io/code-generator/cmd/register-gen/generators/register_external.go

```diff
@@ -0,0 +1,105 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package generators
+
+import (
+	"io"
+	"text/template"
+	"sort"
+
+	clientgentypes "k8s.io/code-generator/cmd/client-gen/types"
+	"k8s.io/gengo/generator"
+	"k8s.io/gengo/types"
+)
+
+type registerExternalGenerator struct {
+	generator.DefaultGen
+	gv              clientgentypes.GroupVersion
+	typesToGenerate []*types.Type
+}
+
+var _ generator.Generator = &registerExternalGenerator{}
+
+func (g *registerExternalGenerator) Filter(_ *generator.Context, _ *types.Type) bool {
+	return false
+}
+
+func (g *registerExternalGenerator) Finalize(context *generator.Context, w io.Writer) error {
+	typesToGenerateOnlyNames := make([]string, len(g.typesToGenerate))
+	for index, typeToGenerate := range g.typesToGenerate {
+		typesToGenerateOnlyNames[index] = typeToGenerate.Name.Name
+	}
+
+	registerTemplateParam := struct {
+		clientgentypes.GroupVersion
+		TypesToGenerate []string
+	}{
+		g.gv,
+		typesToGenerateOnlyNames,
+	}
+
+	// sort the list of types to register, so that the generator produces stable output
+	sort.Strings(registerTemplateParam.TypesToGenerate)
+
+	temp := template.Must(template.New("register-template").Parse(registerExternalTypesTemplate))
+	return temp.Execute(w, registerTemplateParam)
+}
+
+var registerExternalTypesTemplate = `
+// GroupName specifies the group name used to register the objects.
+const GroupName = "{{.Group}}"
+
+// GroupVersion specifies the group and the version used to register the objects.
+var GroupVersion = v1.GroupVersion{Group: GroupName, Version: "{{.Version}}"}
+
+// SchemeGroupVersion is group version used to register these objects
+// Deprecated: use GroupName instead.
+var SchemeGroupVersion = schema.GroupVersion{Group: GroupName, Version: "{{.Version}}"}
+
+// Resource takes an unqualified resource and returns a Group qualified GroupResource
+func Resource(resource string) schema.GroupResource {
+	return SchemeGroupVersion.WithResource(resource).GroupResource()
+}
+
+var (
+	// localSchemeBuilder and AddToScheme will stay in k8s.io/kubernetes.
+	SchemeBuilder      runtime.SchemeBuilder
+	localSchemeBuilder = &SchemeBuilder
+    // Depreciated: use Install instead
+	AddToScheme        = localSchemeBuilder.AddToScheme
+	Install            = localSchemeBuilder.AddToScheme
+)
+
+func init() {
+	// We only register manually written functions here. The registration of the
+	// generated functions takes place in the generated files. The separation
+	// makes the code compile even when the generated files are missing.
+	localSchemeBuilder.Register(addKnownTypes)
+}
+
+// Adds the list of known types to Scheme.
+func addKnownTypes(scheme *runtime.Scheme) error {
+	scheme.AddKnownTypes(SchemeGroupVersion,
+    {{ range .TypesToGenerate -}}
+		&{{.}}{},
+    {{ end -}}
+	)
+    // AddToGroupVersion allows the serialization of client types like ListOptions.
+	v1.AddToGroupVersion(scheme, SchemeGroupVersion)
```

Comment:
> make this dynamic such that the metav1 is automatically imported by the import tracker.

@sttts  I couldn't force the import tracker to import `metav1`, maybe because it is an alias ?

I decided to use `v1` package, is that okay ?

Reply:
Compare https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/code-generator/cmd/client-gen/generators/generator_for_group.go#L91. That way you can dynamically import funcs, without automatic conflict resolution. A hardcoded metav1 is fine as well here as the chance of conflict is low.

Reply:
alright, I think I finally have it.